### PR TITLE
Change test logic

### PR DESCRIPTION
**Pull request by all3fox**
Curently, I am working on the change that is supposed to solve #33. This PR demonstrates that it is possible by simply changing the `ticketPattern` a bit. However, this PR massively changes `test.js`. Here is the reason for it:

Previously, whenever `test.js` tested commitplease against a commit message that required a change in the default settings, that change would be tested against that one message. The `test.js` proposed here allows to first create a number of commitplease configurations and then test them against all (or some subset) of messages (without copy-pasting the messages themselves). The commit message explains the implementation details a bit more.

I am sure there are many things that can be done better, please take a look. As for prohibiting `#nnnn`-style references, `options3` in `test.js` does that and still works fine on all of the messages.

Update: I see that Node 0.12 and Node 4 failed in CI (Node-stable works though). Sorry, I was coding with Node 6 in mind, will fix that.

### CODE REVIEW

**Comment by jzaefferer** [comment]
This looks good to me! I like the new structure, that should scale well as we add more formats and tests as well.

As for pre-node-6 support: We decided against using ES6 features in the module itself, but using them for tests seems okay. You could see if you can make babel or similar tooling work just for the tests, instead of rolling back to ES5.

**Comment by all3fox** [comment]
I have finished changing the logic for tests. I have also added transpiling with babel. However, the ugly way I do it inside `package.json` (creating a temporary transpiled file, running it and deleting it) is setting off the `test-watch` script (it "detects changes" and runs non-stop). Do you perhaps know a better way to do that?

As for prohibiting the hash, it has been unit-tested as profile3 against all messages and works.

File: package.json [hunk 1]
```diff
@@ -25,6 +27,9 @@
   },
   "license": "MIT",
   "devDependencies": {
+    "babel-cli": "^6.10.1",
+    "babel-preset-es2015": "^6.9.0",
+    "babel-preset-stage-0": "^6.5.0",
```
Comment (jzaefferer) [inline package.json]: What stage-0 feature are you using? I (vaguely) prefer to stick with es2015.

File: package.json [hunk 2]
```diff
@@ -3,11 +3,13 @@
...
   "main": "index.js",
-  "bin": {"commitplease": "./commit-msg-hook.js"},
+  "bin": {
+    "commitplease": "./commit-msg-hook.js"
+  },
   "scripts": {
     "install": "node install",
     "uninstall": "node uninstall",
-    "test": "standard && nodeunit test.js",
+    "test": "standard && babel test.js -o tmp.js && nodeunit tmp.js && rm tmp.js",
```
Comment (jzaefferer) [inline package.json]: Let's try to replace this with a direct call to node, while also loading the babel register module, something like this:
```
node -r 'babel-core/register' node_modules/nodeunit/bin/nodeunit test.js
```
Which doesn't work, failing with this:
```
/Users/jzaefferer/dev/jzaefferer/commitplease/node_modules/babel-core/lib/transformation/file/options/option-manager.js:126
      if (!option) this.log.error("Unknown option: " + alias + "." + key, ReferenceError);
                           ^

TypeError: Cannot read property 'error' of undefined
```
Can you give that a try? Might be a simple fix.

**Comment by all3fox** [comment]
`babel-register` indeed looks like the way to go! Now `npm run test-watch` works, however there is a catch: I was not able to do `require('babel-register')({presets: 'es2015'})` at the top of `test.js` --- it was not doing anything. Instead, I had to create a separate `test-babel.js` entry point and `require("./test.js")` from it.

[Some say](http://stackoverflow.com/a/29425761/1269892) it is not possible to transpile the file where you `require("babel-register")` from, [others say](http://stackoverflow.com/a/35002082/1269892) it is possible. I tried doing it without an extra .js file --- did not work for me :(

**Comment by jzaefferer** [comment]
Did you try the approach I suggested? With `-r` you can make node load additional modules, getting around the requirement for an extra file. I don't know how to make that run nodeunit though.

**Comment by all3fox** [comment]
@jzaefferer it worked for me with the following line in `package.json`:
```
"test": "standard && node --require babel-register node_modules/.bin/nodeunit test.js"
```

However, `node 0.12` does not have the `--require` flag, so it did not work after all.

Anyway, there appears to be an option to substitute `node` with a `node-babel` call to a heavier script that is not supposed to be used in production but I guess can be used for testing? It adds on-the-fly transpiling, I pushed it and deleted the extra test file, what do you think?

**Comment by jzaefferer** [comment]
Execellent! I don't think the performance of `babel-node` is a concern here.

**Comment by all3fox** [comment]
Thank you. I am thinking of putting that on npm now:
```
npm version patch
npm publish
```
but I have never done that before. It also looks like I need to tag it in git?

**Comment by jzaefferer** [comment]
`npm version patch` will also create a git tag, you'd still have to push the tag (along with the new commit) to this repo afterwards.

I suggest to make this is a minor release, since we made API changes, hopefully fully compatible with existing installations.

**Comment by all3fox** [comment]
Done:
```
npm version minor
npm push upstream --tags
npm publish
```

**Comment by jzaefferer** [comment]
Nice! Next would a PR for jquery-core to make use of the new option, right?

**Comment by all3fox** [comment]
@jzaefferer yes, here it is: jquery/jquery#3176

**Comment by jzaefferer** [comment]
👍 👍 👍

**Pull request merged**